new_dosi_adaptation.nlogo

```diff
@@ -0,0 +1,725 @@
+      let counter 0
+      let match-counter 0
+      foreach firm-inputs [i ->
+        set counter (counter + 1)
+        if member? i ready-goods[
+          set match-counter (match-counter + 1)
+        ]
+      ]
+      if counter = match-counter [
+        set ready-firm-types lput k ready-firm-types
+      ]
+    ]
+  ]
+  report ready-firm-types
+end
+
+to initiate-framework-agreements ;run by the observer, initiates all of the framework agreements
+  let good-negotiated-for? table:make
+  foreach GOOD-TYPES[g ->
+    ifelse g = 0[
+      table:put good-negotiated-for? g true
+    ][
+      table:put good-negotiated-for? g false
+    ]
+  ]
+  while [any? turtles with [not done-negotiating?]][
+    let ready-firm-types current-negotiating-goods good-negotiated-for?
+    foreach ready-firm-types [f ->
+      let num-firms number-firms f
+      ask turtles with [uses-input? f][
+        let num-negotiations (ceiling (abs (random-normal (num-firms / 10) 1)))
+        let buyer self
+        let buyer-profits current-profits
+        let buyer-preferred-index preferred-index
+        let sum-avg-costs 0
+        ask n-of num-negotiations turtles with [firm-type = f][
+          negotiate-agreement buyer buyer-profits buyer-preferred-index
+          ask link-with buyer [
+            set sum-avg-costs (sum-avg-costs + ((table:get INDICES index) * quantitative-value))
+          ]
+        ]
+        let avg-unit-cost (sum-avg-costs / num-negotiations)
+        let current-input-info table:get input-information f
+        table:put current-input-info "Average unit cost" avg-unit-cost
+        table:put current-input-info "Agreements-negotiated?" true
+        table:put input-information f current-input-info
+      ]
+      table:put good-negotiated-for? f true
+    ]
+  ]
+end
+
+to negotiate-agreement [buyer buyer-profits buyer-preferred-index] ;run by the seller firm, negotiates an individual agreement
+  let total-cost get-atc
+  let seller-profits current-profits
+  let seller-preferred-index preferred-index
+  create-framework-agreement-to buyer [
+    set quantitative-value (total-cost * MARKUP-RULE)
+    (ifelse
+      buyer-profits > seller-profits[
+        set index buyer-preferred-index
+      ]
+      seller-profits > buyer-profits[
+        set index seller-preferred-index
+      ]
+      buyer-profits = seller-profits[
+        ifelse random 2 = 0[
+          set index seller-preferred-index
+        ][
+          set index buyer-preferred-index
+        ]
+      ]
+    )
+    set expiration-date time:plus DATE 5 "years"
+    set purchase-order 0
+    set delivered? false
+  ]
+end
+
+to-report get-atc ;run by a turtle, reports the average total cost
+  let running-total 0
+  foreach table:keys input-information[k ->
+    let current-input table:get input-information k
+    set running-total (running-total + (table:get current-input "Average unit cost"))
+  ]
+  report running-total
```

Comment:
replace this whole thing with `sum map [i -> table:get (table:get input-information i)  "Average unit cost"] table:keys input-information`

Reply:
Not going to change this yet because I think there might be a problem in the way I calculated atc